test/e2e/storage/testsuites/snapshottable_stress.go

```diff
@@ -0,0 +1,233 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This suite tests volume snapshots under stress conditions.
+
+package testsuites
+
+import (
+	"context"
+	"sync"
+
+	"github.com/onsi/ginkgo"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	errors "k8s.io/apimachinery/pkg/util/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2epv "k8s.io/kubernetes/test/e2e/framework/pv"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	e2evolume "k8s.io/kubernetes/test/e2e/framework/volume"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+)
+
+type snapshottableStressTestSuite struct {
+	tsInfo TestSuiteInfo
+}
+
+type snapshottableStressTest struct {
+	config        *PerTestConfig
+	testOptions   StressTestOptions
+	driverCleanup func()
+
+	pods      []*v1.Pod
+	volumes   []*VolumeResource
+	snapshots []*SnapshotResource
+	// Because we are appending snapshot resources in parallel goroutines.
+	snapshotsMutex sync.Mutex
+
+	// Stop and wait for any async routines.
+	ctx    context.Context
+	wg     sync.WaitGroup
+	cancel context.CancelFunc
+}
+
+var _ TestSuite = &snapshottableStressTestSuite{}
+
+// InitSnapshottableStressTestSuite returns snapshottableStressTestSuite that implements TestSuite interface
+func InitSnapshottableStressTestSuite() TestSuite {
+	return &snapshottableStressTestSuite{
+		tsInfo: TestSuiteInfo{
+			Name: "snapshottable-stress",
+			TestPatterns: []testpatterns.TestPattern{
+				testpatterns.DynamicSnapshotDelete,
+				testpatterns.DynamicSnapshotRetain,
+			},
+			SupportedSizeRange: e2evolume.SizeRange{
+				Min: "1Mi",
+			},
+			FeatureTag: "[Feature:VolumeSnapshotDataSource]",
+		},
+	}
+}
+
+func (t *snapshottableStressTestSuite) GetTestSuiteInfo() TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *snapshottableStressTestSuite) SkipRedundantSuite(driver TestDriver, pattern testpatterns.TestPattern) {
+}
+
+func (t *snapshottableStressTestSuite) DefineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	var (
+		driverInfo          *DriverInfo
+		snapshottableDriver SnapshottableTestDriver
+		cs                  clientset.Interface
+		stressTest          *snapshottableStressTest
+	)
+
+	// Check preconditions before setting up namespace via framework below.
+	ginkgo.BeforeEach(func() {
+		driverInfo = driver.GetDriverInfo()
+		if driverInfo.StressTestOptions == nil {
```

Comment:
should we validate the stress test args?

Reply:
Done. Should we skip in this case, or fail? I could verify this and fail it at the end of this function after we've verified the driver is capable of snapshotting.

Reply:
Done